server : vision support via libmtmd

ngxson opened the pull request:
Cont #12849

This is my first trial to bring `libmtmd` to `server.cpp`.

## How to use this?

See this [documentation](https://github.com/ggml-org/llama.cpp/blob/master/docs/multimodal.md)

Web UI now also support uploading image (see demo below)

## Implementation

The main implementation is built around `mtmd.h` and `struct server_tokens` in `utils.hpp`:

The idea is that:
- `struct server_tokens` will make most of the existing text tokens logic to work without too many changes
- `struct server_tokens` can optionally contains image tokens (managed by `libmtmd`)
- `update_slots()` calls `libmtmd` as soon as it sees an image token

```cpp
    // map a **start** position in tokens to the image chunk
    std::unordered_map<llama_pos, mtmd::input_chunk_ptr> map_pos_to_image;

    // list of tokens
    // it can include LLAMA_TOKEN_NULL, which is used to indicate a token that is not a text token
    // a mtmd_input_chunk can occupy multiple tokens, one llama_token per **position**
    // important: for models using mrope, an image can contain multiple tokens but will use only one **position**
    llama_tokens tokens;

    // for ex. with input of 5 text tokens and 2 images:
    //      [0] [1] [2] [3] [4] [img0] [img0] [img0] [img1] [img1]
    // pos  0   1   2   3   4   5      6      7      8      9
    // map_pos_to_image will contain: {5, img0}, {8, img1}
```

## TODOs

- [x] automatically deactivate certain features if vision is enabled, we will work on these features later
- [x] implement hash function for image (to keep track of the cache)
- [x] fix detokenize(server_inp_chunk)
- [x] add more error handlings
- [x] support remote `image_url` in addition of `base64`
- [x] add tinygemma3 model for CI test
- [x] add image upload to web UI
- [x] update docs

## Demo (webui)

The server can be run with this command:

```bash
llama-server -hf ggml-org/gemma-3-4b-it-GGUF
```

Then access the web UI (default at `localhost:8080`):

<img width="1073" alt="image" src="https://github.com/user-attachments/assets/39ee8582-2387-4943-ad5c-4ca64b00c2d0" />

## Demo (API)

<details>
<summary>Click to see</summary>


```py
import json
import base64
from openai import OpenAI

client = OpenAI(base_url="http://127.0.0.1:8080/v1", api_key="sk-test", timeout=9999)

# we support both remove image_url and base64 ; example below is for base64 image (read from disk)

# Function to encode the image
def encode_image(image_path):
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode("utf-8")


# Path to your image
image_path = "../models/bliss.png"

# Getting the Base64 string
base64_image = encode_image(image_path)

response = client.chat.completions.create(
    model="gpt-4o",
    temperature=0.1,
    stream=True,
    messages=[
        {
            "role": "user",
            "content": [
                { "type": "text", "text": "describe what you see in details" },
                {
                    "type": "image_url",
                    "image_url": {
                        # alternatively, you can put the remote image url here, example: "http(s)://....."
                        "url": f"data:image/png;base64,{base64_image}",
                    },
                },
            ],
        }
    ],
)

for chunk in response:
    print(chunk.choices[0].delta.content, end="")

print("\n\n")
```

</details>

ngxson committed:
server : (experimental) vision support via libmtmd

ngxson committed:
mtmd : add more api around mtmd_image_tokens

ngxson committed:
mtmd : add more api around mtmd_image_tokens

ngxson committed:
mtmd : ability to calc image hash

ngxson committed:
shared_ptr for mtmd_image_tokens

ngxson committed:
move hash to user-define ID (fixed)

qnixsynapse commented:
Awesome work. However, I noticed the model usually ignores the text prompt when the prompt is the first in the conversation!
![image](https://github.com/user-attachments/assets/8523ffdd-7a44-41e8-adc3-a41b03a0b1c3)

ngxson commented:
@qnixsynapse can you capture the raw http request? If the json paymoad is big, you can share it via a gist

qnixsynapse commented:
@ngxson Will this be okay? https://gist.github.com/qnixsynapse/a4c61368d05180d3cb6c00f1baedf92c

ngxson commented:
at minimum I ask for this https://wiki.wireshark.org/hyper_text_transfer_protocol

not the raw IP packet

qnixsynapse commented:
I don't have wireshark installed unfortunately. But you can still inspect for example:
POST /v1/chat/completions HTTP/1.1
Host: localhost:8080
Authorization: Bearer -key
Content-Type: application/json
Accept: */*
Accept-Encoding: gzip, deflate
User-Agent: Python/3.11 aiohttp/3.11.11
Content-Length: 615117

And the request data in line 88.

ngxson committed:
Merge branch 'xsn/mtmd_image_api' into xsn/server_mtmd

ngxson committed:
abstract out the batch management

ngxson commented:
@qnixsynapse I had a problem with my logic, which make it discard the text batch comes before the image batch.

It should be fixed now, could you give it a try?

ngxson commented:
Btw @ggerganov I'm noting here for visibility: while working on this PR, I realize that I can have 2 refactoring which can be done in their dedicated PR:
- The first one is quite simple, currently `server_task` is passed-by-copy in some places, we need to add some `std::move`
- The second one is a bit more tricky. Currently, we track everything using a `std::vector<llama_token>`. However, for multimodal, I introduced the notion of "input chunks" along with `libmtmd`. Server need to be adapted to work with chunks of tokens / embeddings instead of a simple list of tokens.  
  In the current PR, I'm kinda hacking this by having `server_inp_chunk` to wrap around one single text token (so most of the text-related logic are unchanged). But obviously this brings some complication when dealing with both text + image chunks. Do you have any better ideas?

And I also have a question regarding the logic around `batch_view`. IIRC, this is because sometimes the batch is too large for `llama_decode` to process, so we may want to reduce the input batch size (dynamically). However, we also internally split the batch into `ubatch`, so I'm wondering if this logic is now obsolete.

ggerganov commented:
> And I also have a question regarding the logic around batch_view. IIRC, this is because sometimes the batch is too large for llama_decode to process, so we may want to reduce the input batch size (dynamically). However, we also internally split the batch into ubatch, so I'm wondering if this logic is now obsolete.

This was useful mainly before the defragmentation support was added. The reason is that with time the KV cache can become highly fragmented and even if it has capacity for `n_tokens` it won't be able to find a contiguous slot, so attempting to split the batch into smaller chunks was a way to workaround this. With defragmentation enabled by default this is now rarely necessary. So yes, this should be simplified in a separate PR.

I'll think about the input chunk question today and let you know if I have any thoughts.

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

qnixsynapse commented on common/arg.cpp:
```diff
@@ -833,9 +833,11 @@ static bool common_params_parse_ex(int argc, char ** argv, common_params_context
         params.mmproj.hf_repo = params.model.hf_repo;
     }
+    // TODO @ngxson : this will break non-vision model with -hf, need to fix before merging
     common_params_handle_model(params.mmproj,            params.hf_token, "", true);
```
@ngxson Is it possible to add a `--no-offload-mmproj` param here to keep the mmproj model on the CPU and the larger text model on GPU?

We can use mtmd_param_param: use_gpu=false to keep the projector model on CPU itself.

https://github.com/ggml-org/llama.cpp/blob/0019279bb56f028e4eee19b59b19750736c719f7/examples/llava/mtmd.h#L52-L58

It will be useful where the GPU VRAM is limited.

ngxson replied:
It's added in https://github.com/ggml-org/llama.cpp/pull/13093

the flag is `--no-mmproj-offload` and not `--no-offload-mmproj`, to align with the existing `--no-kv-offload`

qnixsynapse replied:
Thanks alot!!!

Beinsezii commented:
Seems like the batch decoding dies when you send a variety of longer requests.

`common/common.cpp:1159: GGML_ASSERT(batch.seq_id[batch.n_tokens] && "llama_batch size exceeded") failed`

Easiest way to trigger is to just wiggle the sequence length around, like with the example code
```py
import json
import base64
from openai import OpenAI

client = OpenAI(base_url="http://127.0.0.1:8080/v1", api_key="sk-test", timeout=9999)

# Function to encode the image
def encode_image(image_path):
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode("utf-8")


# Path to your image
image_path = "../models/bliss.png"

# Getting the Base64 string
base64_image = encode_image(image_path)

for mult in [100, 200]:  # (beinsezii) make sure it has to rebuild some cache the 2nd time
    response = client.chat.completions.create(
        model="gpt-4o",
        temperature=0.1,
        stream=True,
        messages=[
            {
                "role": "user",
                "content": [
                    { "type": "text", "text": "describe what you see in details\n" * mult },
                    {
                        "type": "image_url",
                        "image_url": {
                            "url": f"data:image/png;base64,{base64_image}",
                        },
                    },
                ],
            }
        ],
    )

    for chunk in response:
        print(chunk.choices[0].delta.content, end="")

    print("\n\n")
```

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
small fix

ngxson commented on examples/server/utils.hpp:
```diff
@@ -913,3 +957,285 @@ static std::vector<common_adapter_lora_info> parse_lora_request(
+
+    void reserve_embd_batch(float * embd, int32_t n_tokens, llama_pos pos_0, llama_seq_id seq_id) {
+        GGML_ASSERT(n_tokens <= (int32_t)pos.size());
+        seq_ids[n_tokens] = nullptr;
```
@Beinsezii yes seems like it's due to this single line which shouldn't be here, I'm removing this, will push a commit

ngxson replied:
Should be fixed in https://github.com/ggml-org/llama.cpp/pull/12898/commits/f8bc46629fa4e669697106f251dd9e7d957ff218

And btw, as https://github.com/ggml-org/llama.cpp/pull/13012 is merged, we can now test other models (except for qwen2vl, not yet supported)

Beinsezii replied:
Oh nice, server will automatically have everything libmtmd does? Like https://github.com/ggml-org/llama.cpp/pull/13050

I'll try to break the decode again here in a bit.

ngxson replied:
yes, everything supported by libmtmd will be available on server

Beinsezii replied:
Using gemma 27 I've run multiple fat images interspersed with varying text sequences and it seems to be stable now. I'm up to like 5 images in one prompt, so that should be all good now.

I know from experience GLM is broken on ROCm, but I am curious about MiniCPM and SmolVLM so I'll try those at some point.

Beinsezii replied:
IBM granite fails to encode when feeding in a 3840x2160 image but otherwise it's been a positive experience.

ngxson replied:
yes some models requires larger `--batch` for now. this will be fixed soon

ngxson replied:
larger --batch is no longer required. for some models, you may want to bump the ctx size to 8k or 16k because they use a lot more tokens for bigger images

ngxson committed:
refactor logic adding tokens to batch

ngxson committed:
implement hashing image

ngxson commented:
Image hashing (SHA1) is implemented in https://github.com/ggml-org/llama.cpp/pull/12898/commits/f5420e1d90bf7228c12bb5f8cd85808c4cb00ba8 , which should allow reusing KV cache for image tokens.

Would be nice if anyone can test this (or even better, write a python script to hammer this)

Beinsezii commented:
> Image hash (SHA1) is implemented in [f5420e1](https://github.com/ggml-org/llama.cpp/commit/f5420e1d90bf7228c12bb5f8cd85808c4cb00ba8) , which should allow reusing KV cache for image tokens.

Is this actually implemented? Even using your own example with `bliss.png` it encodes every time as of f5420e1d90bf7228c12bb5f8cd85808c4cb00ba8

ggerganov commented on examples/server/CMakeLists.txt:
```diff
@@ -34,8 +34,10 @@ endforeach()
 install(TARGETS ${TARGET} RUNTIME)
 
+target_include_directories(${TARGET} PRIVATE ../llava)
+target_include_directories(${TARGET} PRIVATE ../gguf-hash/deps/sha1) # TODO @ngxson : this is a hacky way to get this working, to be fixed before merging
```
A basic FNV hash should be good enough:

https://github.com/ggml-org/llama.cpp/blob/1d735c0b4fa0551c51c2f4ac888dd9a01f447985/ggml/src/ggml-rpc/ggml-rpc.cpp#L210-L220

99991 replied:
If you process the entire Laion-5B dataset with a 64 bit hash function, you have got a 50 % chance for a hash collision.

```python
# https://en.wikipedia.org/wiki/Birthday_problem#Approximations
import math

n = 5 * 10**9
number_of_birthdays = 2**64

p_no_collision = math.exp(-n * (n - 1) / (2 * number_of_birthdays))

print("Collision probability:", 1 - p_no_collision)
```

With SHA-1 (160 bits), the probability would be approximately 0. But to be safe, we should use SHA-256 instead, for which no collisions are known. I don't think that 5ms are worth the risk.

ngxson replied:
> If you process the entire Laion-5B dataset with a 64 bit hash function, you have got a 50 % chance for a hash collision.

Do you have the python code to calculate FNV-1 hash (or other hash functions) from a given dataset? I want to try myself on a colab notebook just to see how it works

I don't think our problem is performance, as the hash is calculate at HTTP thread. The more important question is how to have the least dependency for llama-server

99991 replied:
> Do you have the python code to calculate FNV-1 hash (or other hash functions) from a given dataset? I want to try myself on a colab notebook just to see how it works

You can simply wrap the C function from above with `ctypes`. The script below will find image files with the same hash in the given directory.

```python
import os, ctypes, collections, multiprocessing.pool
from pathlib import Path

directory = "."

filetypes = [".jpg", ".jpeg", ".png", ".gif", ".bmp", ".tiff", ".webp"]

paths = (path for path in Path(directory).rglob("*") if path.is_file() and path.suffix.lower() in filetypes)

# Compile fnv1a hash function
Path("source.c").write_text("""
#include <stdint.h>
#include <stddef.h>

uint64_t fnv1a(const uint8_t * data, size_t len){
    const uint64_t fnv_prime = 0x100000001b3ULL;
    uint64_t hash = 0xcbf29ce484222325ULL;

    for (size_t i = 0; i < len; ++i) {
        hash ^= data[i];
        hash *= fnv_prime;
    }
    return hash;
}
""")

assert 0 == os.system("gcc -shared -o fnv.so -fPIC source.c")

os.remove("source.c")

dll = ctypes.CDLL("./fnv.so")
dll.fnv1a.argtypes = [ctypes.POINTER(ctypes.c_char), ctypes.c_size_t]
dll.fnv1a.restype = ctypes.c_uint64

def fnv1a(x: bytes) -> int:
    return dll.fnv1a(x, len(x))

# Testing correctness with some hash collision examples
# Thanks to Rudi Cilibrasi http://www.isthe.com/chongo/tech/comp/fnv/
assert fnv1a(bytes([0xd5, 0x6b, 0xb9, 0x53, 0x42, 0x87, 0x08, 0x36])) == 0
assert fnv1a(bytes([0x07, 0x1e, 0x62, 0x37, 0x2c, 0x02, 0x40, 0x42, 0x14, 0x69])) == 0

def hash_file(path):
    size = path.stat().st_size
    data = path.read_bytes()
    hash_value = fnv1a(data)
    hash_value = int.to_bytes(hash_value, 8, 'big')
    return hash_value, size, str(path)

files_with_same_hash = collections.defaultdict(list)

# Find hash collisions in parallel
with multiprocessing.pool.ThreadPool() as pool:
    for i, result in enumerate(pool.imap_unordered(hash_file, paths)):
        hash_value, size, path = result

        files = files_with_same_hash[hash_value]

        for other in files:
            if Path(other).read_bytes() != Path(path).read_bytes():
                print(f"\n\nFound hash collision {hash_value.hex()}: {path}, {other}\n")

        files.append(path)

        if len(path) > 40: path = "..." + path[-40:]
        print(f"{i + 1}: {hash_value.hex()}, {path}", end="\r")
```

ngxson replied:
> We should also consider the possibility of hand-crafted collisions. If, for example, someone was using a VLM-based filter, it could be circumvented by first submitting a harmless image and then submitting a violating image with the same FNV-1 hash.

This is an interesting attack vector. I think it's quite interesting in our case, as we are hashing the raw file instead of the bitmap. An attacker can reproduce such a collisions by adding hand-crafted metadata to the image. I think the safer solution would be to hash the bitmap instead.

ngxson commented:
> Is this actually implemented, or is it just a framework? Even using your own example with `bliss.png` it encodes every time as of [f5420e1](https://github.com/ggml-org/llama.cpp/commit/f5420e1d90bf7228c12bb5f8cd85808c4cb00ba8)

What is your test code or request? I rerun the python test code in the PR description more than once, and it does not re-encoding the image

Beinsezii commented:
> What is your test code or request? I rerun the python test code in the PR description more than once, and it does not re-encoding the image

Using your demo code with bliss.png I get `image encoded in 400ms` every single time I run the script, even with 100% identical requests. I'll do some picking see why mine does that. I ran cmake again just to confirm I'm on the latest commit.

ngxson commented:
@Beinsezii Hmm ok thanks for spotting that. It's not a race condition, but without `SHA1Init` I think the initial vector is initialized to a random value on heap/stack

Beinsezii commented:
> @Beinsezii Hmm ok thanks for spotting that. It's not a race condition, but without `SHA1Init` I think the initial vector is initialized to a random value on heap/stack

that makes sense but now im even more confused as to why it was consistent on your end lol. im not even sure zeroed pages would work because surely enough stuff happens over multiple requests it would use reclaimed memory. maybe your discrete rng lava lamps are unplugged

andportnoy commented:
A hash can be an arbitrary byte sequence, right? It's not necessarily a valid string. You probably want to print it out byte by byte, using something like:

```c++
printf("hash = ");
for (int i = 0, n = sizeof result; i < n; ++i) {
  printf("%02hhx", result[i]);
}
printf("\n");
```

ngxson commented:
> A hash can be an arbitrary byte sequence, right? It's not necessarily a valid string.

Yes but storing it hex string is easier for debugging, so it must be converted to a hex string to prevent potential problems null byte. This conversion is currently missing in the code.

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
use FNV hash, now hash bitmap instead of file data

ngxson committed:
allow decoding image embedding to be split into batches

ngxson committed:
rm whitespace

ngxson commented:
Significant changes in last commits:
- bump to latest `master`, we're now supporting Pixtral 12B
- using FNV hash, with the hash data is the image bitmap (NOT the raw file data)
- support large image batches, so models like granite-vision or minicpm-v won't crash

Beinsezii commented:
> bump to latest `master`, we're now supporting Pixtral 12B

curious if small 3.1 is the same vision mechanism or if that will need more work as well

ngxson commented:
> Update: seems like Pixtral is broke

Which backend you're using? Does it give the same result when running via `llama-mtmd-cli`?

Beinsezii commented:
> Which backend you're using? Does it give the same result when running via `llama-mtmd-cli`?

ROCm and it seems to be temperature dependent?

Like 0.1 temp it will reply
`It seems we're starting with an image of a serene landscape featuring a clear blue sky transitioning into lush green fields below.`
where temp 1.0 is
`it seems that the image you've shared contains a pattern of repeating colors and shapes that might be difficult to describe precisely without more context.`

Meanwhile on CPU it always recognizes it as a landscape even at temp 2.0. ROCm at 2.0 claims there isn't an image at all lol. I imagine something is wrong because I don't think temp should swing the results that hard for such a simple prompt.

Haven't tried Vulkan yet. Identical behavior with mtmd-cli. Shall I open an issue?

HAV0X1014 commented:
I'm getting wildly incorrect outputs with Pixtral. I'm using the server API and llama-mtmd-cli and the server seems to completely ignore that I've sent an image, but the CLI outputs garbage - mentioning either a mosaic of colors or just outputting complete nonsense. This image in particular made it go nuts, counting up from 2013 until generation stopped.
![IMG_20240224_194019](https://github.com/user-attachments/assets/f5cf9809-763a-424a-abec-7388db2e1f29)

I'm using a 7900xtx, compiled with ROCm. Running it on CPU and GPU produced different, but still incorrect, results.

Beinsezii commented:
@HAV0X1014 if you're trying cpu, try a clean cpu only build without HIP compiled at all. For some reason compiling with HIP but using `--ngl 0` can still break some models. GLM 4 is the same way.

Beinsezii commented:
Is there a way to pass images via non-chat completion yet? I see in the [server readme](https://github.com/ggml-org/llama.cpp/blob/master/examples/server/README.md) at one point `/completions` could substitute images like
```nu
http post http://127.0.0.1:8080/completion --content-type application/json {
    prompt: 'What is in this image?[img-12]',
    "image_data": [{"data": (open /tmp/bliss.png | encode base64), "id": 12}]
}
```
but I don't believe that's functional anymore.

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
disable some features when mtmd is on

ngxson commented:
Putting pixtral aside (because the bug is not related to server support), would you be interested in testing other models?

I think the current PR is in a "working" state, meaning it can already handle the most important use case, the `/chat/completions` endpoint. That's the first phase on my plan. I'm thinking about moving to the second phase, it's to iterate on the current idea and fix bugs.

So would be nice to know if the current approach works well with these cases:
- Chat completion, text-only
- Chat completion, text and image
- Chat completion, text and image, but multiple texts and multiple images interleaved in the same message
- Chat completion, with KV reuse

Beinsezii commented:
Ah, seems like `-hf` might not work properly on models without an mmproj? I did `bin/llama-server -hf bartowski/THUDM_GLM-4-32B-0414-GGUF:Q4_K_L` and it 404'd repeatedly until I switched to the non-mtmd server branch.

ngxson commented:
> The only real annoyance I've had monkeying around so far is the mmproj being locked to the GPU eating half of the memory I normally use for k/v cache.

There is a recently added `--no-mmproj-offload`, I don't add fine control for offloading layers because the mmproj usually very small compared to text model, and also it may take me more time to actually implement it.

> a cache hit when the first X of a new request matches in tokens so it doesn't re-process?

yes it is

> Ah, seems like `-hf` might not work properly on models without an mmproj?

it should be fixed in https://github.com/ggml-org/llama.cpp/pull/13082 , but I haven't tested yet. Are you running the latest commit in the current PR? https://github.com/ggml-org/llama.cpp/pull/12898/commits/2df8c1a4b422fbf30c01ae1de75506826dd3499f

Beinsezii commented:
> There is a recently added `--no-mmproj-offload`, I don't add fine control for offloading layers because the mmproj usually very small compared to text model, and also it may take me more time to actually implement it.

Ah, that's perfect. I missed that on the new commit.

ngxson committed:
fix --no-mmproj-offload

ngxson commented:
ok thanks for reporting, --no-mmproj-offload should be fixed in the last commit

Beinsezii commented:
> ok thanks for reporting, --no-mmproj-offload should be fixed in the last commit

confirm it works now. I got all my vram back at the cost of waiting a whole minute for all 7 images to encode lol. 

Currently it seems like the image hashing doesn't actually save the tokens, rather just is a global kv re-use check. I'm on a 7900X, but I imagine someone with a quad core waiting 30 seconds for their image, realizing their prompt had a typo, then waiting 30 more seconds because the history changed so the cache was evicted. Not sure if it's possible to also cache the image tokens themselves or not, but it would probably help some people.

ngxson commented:
yeah I was thinking about caching image embeddings too, but it can be a bit tricky because we don't yet have a cache eviction strategy. but this can be added later on

ngxson committed:
mtmd_context_params no timings

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
refactor server_inp to server_tokens

ngxson commented on examples/server/utils.hpp:
```diff
@@ -913,3 +957,335 @@ static std::vector<common_adapter_lora_info> parse_lora_request(
+// (may need to refactor in near future)
+//
+
+// each chunk can contain either one SINGLE text token or pointer to image
+// this is to simplify the logic of KV cache management
+struct server_token {
+    llama_token txt;
+    std::shared_ptr<mtmd_image_tokens> img;
```
In order not to affect too much on the existing functionalities, I end up using this struct. My idea is that each instance of this struct correspond to exactly 1 token in KV cache.

For functionalities that rely on `std::vector<llama_token>`, like ctx shift or speculative decoding, it can continue to function with minimal change. Ofc we make sure to disable these features if mtmd is used, because these features cannot (yet?) handle image tokens.

Also, here I have to use `shared_ptr` because an image can be represented by multiple tokens, so all of these tokens will point to one single image. I do this way so `size()` automatically reflects the token count in the prompt. It may sound ugly, but I think this is the best interim solution before we come up with something more clean.

@ggerganov If you have a bit of time, could you do a quick review just to see if my global implementation is ok?

ngxson committed:
fix the failing test case

ngxson committed:
init

ngxson committed:
wip

ngxson committed:
Merge branch 'master' into xsn/mtmd_c_api

ngxson committed:
working version

ngxson committed:
add mtmd::bitmaps

ngxson committed:
add test target

ngxson committed:
rm redundant define

ngxson committed:
test: mtmd_input_chunks_free

ngxson committed:
rm outdated comment

ggerganov commented on examples/server/utils.hpp:
```diff
@@ -913,3 +960,335 @@ static std::vector<common_adapter_lora_info> parse_lora_request(
+ * each token can be either a text token or a pointer to an image.
+ * if image usually contains multiple tokens, each token contains a shared_ptr to the same image.
+ */
+struct server_tokens {
+    bool has_mtmd = false;
+    std::vector<server_token> values;
+
```
I think you can maintain `llama_tokens values_text;` in parallel:

```c++
struct server_tokens {
    bool has_mtmd = false;
    std::vector<server_token> values;
    llama_tokens values_text; // <-- keep in sync with values

...
```

This way you would avoid a lot of copies and memory allocations when converting from `server_tokens` to `llama_tokens`.

ngxson replied:
Hmm your suggestion is a bit like what I'm currently doing in https://github.com/ngxson/wllama/pull/173 , what I did is to use all negative numbers from `-1000` (`count--` each time) to mark the image placement.

Not sure if this looks right to you, but another option could be to simply use `LLAMA_TOKEN_NULL` to mark the image token, then have a mapping `llama_pos --> image` to keep track of the image placement. This way, we don't need the `count--` hack that I mentioned above.

ggerganov replied:
This also works functionally, though you would still need to have the `llama_tokens server_tokens::get_text_tokens()` method.

I think ideally you want to avoid this method so that `server_tokens.values_text` could be used interchangeably with the legacy `llama_tokens`. It will not have image tokens or placeholders - just the text tokens. The goal is when multi-modality is not used, to have the same logic in the code as before and also to avoid copies of the tokens on each iteration (for example during speculative decoding).

ggerganov replied:
> Not sure if this looks right to you, but another option could be to simply use LLAMA_TOKEN_NULL to mark the image token, then have a mapping llama_pos --> image to keep track of the image placement. This way, we don't need the count-- hack that I mentioned above.

Yes, actually this should work too and you don't need `get_text_tokens()`. Maybe this is the better option.

ngxson committed:
Merge branch 'master' into xsn/mtmd_c_api

ngxson committed:
fix merging issue

ngxson committed:
explicitly create mtmd::input_chunks

ngxson committed:
mtmd_input_chunk_copy

ngxson committed:
add clone()

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
improve server_input struct

ngxson committed:
clip :  fix confused naming ffn_up and ffn_down

ngxson committed:
rm ffn_i/o/g naming

ngxson committed:
rename n_embd, n_ff

ngxson committed:
small fix

ngxson committed:
Merge branch 'master' into xsn/clip_ffn_up_down_fix

ngxson committed:
no check n_ff

ngxson committed:
Merge branch 'xsn/clip_ffn_up_down_fix' into xsn/server_mtmd

ngxson committed:
fix detokenize

ngxson commented on tools/server/utils.hpp:
```diff
@@ -935,3 +982,277 @@ static std::vector<common_adapter_lora_info> parse_lora_request(
+
+private: // disallow accessing these members directly, risking out-of-sync
+
+    // map a **start** position in tokens to the image chunk
+    std::unordered_map<llama_pos, mtmd_input_chunk> map_pos_to_image;
+
+    // list of tokens
+    // it can include LLAMA_TOKEN_NULL, which is used to indicate a token that is not a text token
+    // a mtmd_input_chunk can occupy multiple tokens, one llama_token per **position**
+    // important: for models using mrope, an image can contain multiple tokens but will use only one **position**
+    llama_tokens tokens;
+
+    // for ex. with input of 5 text tokens and 2 images:
+    //      [0] [1] [2] [3] [4] [img0] [img0] [img0] [img1] [img1]
+    // pos  0   1   2   3   4   5      6      7      8      9
+    // map_pos_to_image will contain: {5, img0}, {8, img1}
```
@ggerganov This is the new way that I propose. To address the risk of desync between `tokens` and `map_pos_to_image`, these members are private.

To provide access to the list of underlaying `tokens`, I added a function `get_text_tokens()`. It will still be a copy in case the caller wants to manipulate the list, an example can be found in the logic of context shift and prompt truncation (tbh I think we should re move the prompt truncation altogether). However, I think these copies can easily be removed by adding functions like `server_tokens::remove_first(n)`

ngxson committed:
Merge branch 'master' into xsn/mtmd_c_api

ngxson committed:
add const to various places

ngxson committed:
add warning about breaking changes

ngxson committed:
Merge branch 'xsn/mtmd_c_api' into xsn/server_mtmd

ngxson committed:
add c api

ngxson committed:
helper: use mtmd_image_tokens_get_n_pos

ngxson committed:
Merge branch 'xsn/mtmd_c_api' into xsn/server_mtmd

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
fix ctx_shift

ngxson committed:
fix name shadowing

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
more strict condition

ngxson committed:
support remote image_url

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
remote image_url log

ngxson committed:
add CI test

ngxson commented:
Most of the important items in my TODO list are now done, I think this PR is pretty much up for testing & reviewing now.

If you're interested, feel free to give it a try and leave your feedback here. Here is a list of example models:

```sh
# Gemma 3
llama-server -hf ggml-org/gemma-3-4b-it-GGUF
llama-server -hf ggml-org/gemma-3-12b-it-GGUF
llama-server -hf ggml-org/gemma-3-27b-it-GGUF

# SmolVLM
llama-server -hf ggml-org/SmolVLM-Instruct-GGUF
llama-server -hf ggml-org/SmolVLM-256M-Instruct-GGUF
llama-server -hf ggml-org/SmolVLM-500M-Instruct-GGUF
llama-server -hf ggml-org/SmolVLM2-2.2B-Instruct-GGUF
llama-server -hf ggml-org/SmolVLM2-256M-Video-Instruct-GGUF
llama-server -hf ggml-org/SmolVLM2-500M-Video-Instruct-GGUF

# Pixtral 12B
llama-server -hf ggml-org/pixtral-12b-GGUF

# Qwen 2 VL
llama-server -hf ggml-org/Qwen2-VL-2B-Instruct-GGUF
llama-server -hf ggml-org/Qwen2-VL-7B-Instruct-GGUF

# Qwen 2.5 VL
llama-server -hf ggml-org/Qwen2.5-VL-3B-Instruct-GGUF
llama-server -hf ggml-org/Qwen2.5-VL-7B-Instruct-GGUF
llama-server -hf ggml-org/Qwen2.5-VL-32B-Instruct-GGUF
llama-server -hf ggml-org/Qwen2.5-VL-72B-Instruct-GGUF

# Mistral Small 3.1 24B (IQ2_M quantization)
llama-server -hf ggml-org/Mistral-Small-3.1-24B-Instruct-2503-GGUF --chat-template mistral-v7
```

For the test client code, see in the description of this PR

ngxson commented on tools/server/tests/utils.py:
```diff
@@ -379,6 +382,21 @@ def jina_reranker_tiny() -> ServerProcess:
+    def tinygemma3() -> ServerProcess:
+        server = ServerProcess()
+        # mmproj is already provided by HF registry API
+        server.model_hf_repo = "ggml-org/tinygemma3-GGUF"
+        server.model_hf_file = "tinygemma3-Q8_0.gguf"
+        server.mmproj_url = "https://huggingface.co/ggml-org/tinygemma3-GGUF/resolve/main/mmproj-tinygemma3.gguf"
```
I trained this tiny model for using in CI, it's just 38M params

The dataset used for training was CIFAR-10 ; however, the quality is quite bad. If know how to train tiny models (less than 50M params), please reach out!

ngxson committed:
do not log base64

Beinsezii commented:
Interestingly small 3.1 gives `llama.cpp/ggml/src/ggml-cpu/ops.cpp:5871: fatal error` for BF16 mmproj but not F16, using `--no-mmproj-offload` since 9GiB is insane.

ngxson commented:
> Pan&Scan in Gemma 3 isn't implemented, right?

We don't implement it because pan&scan involve some complicated prompt engineering (and I doubt if any other runtime support this). It's basically like tool calling.

> Interestingly small 3.1 gives `llama.cpp/ggml/src/ggml-cpu/ops.cpp:5871: fatal error` for BF16 mmproj but not F16, using `--no-mmproj-offload` since 9GiB is insane

I think BF16 is not yet supported everywhere

Re. mem usage, it will be resolve soon when flash attention comes to clip

stduhpf commented:
> 1. will it introduce new args to run the server? didn't see any update to tools/server/README.md
> 2. we will need to actually run 2 models on the same server? sorry if its stupid question, the demo show  only -hf to select single model so i'm a bit confused about the mmproj.

Yes, the added args are `--mmproj` , "--mmproj-url`, `--no-mmproj`,  and `--no-mmproj-offload`.

ngxson commented:
`--help`

<img width="939" alt="image" src="https://github.com/user-attachments/assets/3162992b-71cb-46b0-92b8-a9a420a52dca" />

ggerganov commented on tools/server/utils.hpp:
```diff
@@ -935,3 +1011,271 @@ static std::vector<common_adapter_lora_info> parse_lora_request(
+        tokens.clear();
+    }
+
+    void resize(size_t n) {
+        GGML_ASSERT(n <= tokens.size());
+        // we throw an error if we try to remove a token in the middle of an image
+        // for ex. with input of 5 text tokens and 2 images:
+        //    [0] [1] [2] [3] [4] [img0] [img0] [img0] [img1] [img1]
+        // n  1   2   3   4   5   6      7      8      9      10
+        // allowed to resize      ^                    ^
+        // disallowed to resize          ^      ^             ^
+        if (n > 0) {
+            llama_token last_token = tokens[n - 1];
+            // make sure we never remove tokens in the middle of an image
+            if (last_token == LLAMA_TOKEN_NULL) {
+                find_chunk(n - 1); // will throw an error if the token is not begin-of-chunk
+            }
+        }
+        tokens.resize(n);
+    }
```
Should we remove entires from `map_pos_to_image` here?

ngxson replied:
Fixed in https://github.com/ggml-org/llama.cpp/pull/12898/commits/eeda075ede16418f29c94ea395cead8dcfd025ae

Re. the problem with speculative decoding, I suspect that this can be because of the `cache_tokens.resize()` which check for the image even when server doesn't have multimodal, but I'm not 100% sure. In included a check `if (has_mtmd)` in the `resize()`, could you check if this resolves the issue with speculative?

ngxson replied:
Also this which may improve the performance a bit: https://github.com/ggml-org/llama.cpp/pull/12898/commits/bef122e1a5327fc1d3ede6488cd76bd5176605a1

ggerganov commented:
I think speculative decoding is less-performant with this PR, so there is probably a bug somewhere. Testing with Qwen2.5: `Write the numbers from 1 to 300.`

ngxson committed:
add "has_multimodal" to /props

ngxson committed:
remove dangling image

ngxson committed:
speculative: use slot.cache_tokens.insert

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson commented:
This should now work out of the box with the default Web UI

<img width="1073" alt="Screenshot 2025-05-08 at 15 43 20" src="https://github.com/user-attachments/assets/44c63cc8-afba-470e-b6ed-f81ef08be5d8" />

vgrunner4v commented:
maybe stupid question (sorry!)
so in order to activate it, we should provide --mmproj (in loading local gguf scenario) and if we not include it, then it will consider it as 'multimodel disabled'? asking because of some features that disabled once multimodal is on, like context shift.

pnb commented:
I tried Mistral Small 2504 (Q4_K_M) from `bartowski/mistralai_Mistral-Small-3.1-24B-Instruct-2503-GGUF` and it works terrifically with --no-mmproj-offload or a Q8_0 quant of the mmproj and only 2K context (on a 7900XTX).

It is even able to pass the strawberry test (occasionally):
![Strawberries and blueberries on a plate, with some of the strawberries arranged in the shape of the capital letter R](https://github.com/user-attachments/assets/bb1477fe-f0c7-4938-984d-29614e91d217)

Twice I got Jinja parsing errors with `--jinja`, although I suspect this is unrelated to the image part. I couldn't consistently replicate it but will keep trying in case it is somehow related to libmtmd.

ngxson commented:
For mistral small, it does not come with a default chat template, make sure to specify one, otherwise it will use default chatml which is incorrect

p1-0tr commented:
I experimented with this PR a bit, mostly using gemma3 12b (`bartowski/google_gemma-3-12b-it-GGUF`). HW: mac m4 max. Everything worked great :)

Excited to see this land. Thanks :)

ggerganov commented on tools/server/server.cpp:
```diff
@@ -3380,7 +3491,7 @@ struct server_context {
 
                 slot.cache_tokens.push_back(id);
-                slot.cache_tokens.insert(slot.cache_tokens.end(), ids.begin(), ids.end() - 1);
+                slot.cache_tokens.insert(ids);
```
This fixes the speculative decoding performance:

```suggestion
                slot.cache_tokens.insert({ids.begin(), ids.end() - 1});
```

ggerganov commented on tools/server/server.cpp:
```diff
@@ -2096,16 +2135,11 @@ struct server_context {
         const int32_t n_vocab = llama_vocab_n_tokens(vocab);
-        for (const auto & token : tokens) {
-            if (token < 0 || token >= n_vocab) {
-                return false;
-            }
-        }
-        return true;
+        return inp.validate(n_vocab);
     }
```
I would simplify this with the following in order to keep the validation logic localized in `server_tokens`:

```diff
diff --git a/tools/server/server.cpp b/tools/server/server.cpp
index 9c81cb377..83e93be6c 100644
--- a/tools/server/server.cpp
+++ b/tools/server/server.cpp
@@ -2135,13 +2135,6 @@ struct server_context {
         return ret;
     }
 
-    bool can_be_detokenized(const struct llama_context * ctx, const server_tokens & inp) {
-        const llama_model * model = llama_get_model(ctx);
-        const llama_vocab * vocab = llama_model_get_vocab(model);
-        const int32_t n_vocab = llama_vocab_n_tokens(vocab);
-        return inp.validate(n_vocab);
-    }
-
     bool launch_slot_with_task(server_slot & slot, server_task && task) {
         slot.reset();
         slot.id_task       = task.id;
@@ -2156,8 +2149,7 @@ struct server_context {
             slot.lora = slot.params.lora;
         }
 
-        bool can_detokenize = can_be_detokenized(ctx, slot.prompt_tokens);
-        if (!can_detokenize) {
+        if (!slot.prompt_tokens.validate(ctx)) {
             send_error(task, "Prompt contains invalid tokens", ERROR_TYPE_INVALID_REQUEST);
             return false;
         }
diff --git a/tools/server/utils.hpp b/tools/server/utils.hpp
index 53f72aea1..23163f4fe 100644
--- a/tools/server/utils.hpp
+++ b/tools/server/utils.hpp
@@ -1227,7 +1227,11 @@ public:
     }
 
     // make sure all text tokens are within the vocab range
-    bool validate(llama_token max_vocab_id) const {
+    bool validate(const struct llama_context * ctx) const {
+        const llama_model * model = llama_get_model(ctx);
+        const llama_vocab * vocab = llama_model_get_vocab(model);
+        const int32_t n_vocab = llama_vocab_n_tokens(vocab);
+
         for (size_t i = 0; i < tokens.size(); ++i) {
             auto & t = tokens[i];
             if (t == LLAMA_TOKEN_NULL) {
@@ -1239,7 +1243,7 @@ public:
                 } catch (const std::exception & e) {
                     return false;
                 }
-            } else if (t < 0 || t >= max_vocab_id) {
+            } else if (t < 0 || t >= n_vocab) {
                 return false;
             }
         }
```

ngxson replied:
Thanks! I applied this patch in https://github.com/ggml-org/llama.cpp/pull/12898/commits/f10fc5613d4f9513bef2caea11fc7a722de47b31

ggerganov commented on tools/server/server.cpp:
```diff
@@ -3969,15 +4082,68 @@ int main(int argc, char ** argv) {
+                                                    &inp_txt,
+                                                    bitmaps_c_ptr.data(),
+                                                    bitmaps_c_ptr.size());
+                if (tokenized != 0) {
+                    throw std::runtime_error("Failed to tokenize prompt");
+                }
+                server_tokens tmp(chunks, true);
+                inputs.push_back(std::move(tmp));
+
+            } else {
```
```suggestion
                if (tokenized != 0) {
                    throw std::runtime_error("Failed to tokenize prompt");
                }

                server_tokens tmp(chunks, true);
                inputs.push_back(std::move(tmp));
            } else {
```

ggerganov commented on tools/server/server.cpp:
```diff
@@ -2734,6 +2777,7 @@ struct server_context {
                 } break;
             case SERVER_TASK_TYPE_SLOT_SAVE:
                 {
+                    if (!ensure_no_mtmd(task.id)) break;
```
nit:

```suggestion
                    if (!ensure_no_mtmd(task.id)) {
                        break;
                    }

```

ggerganov reviewed (approved):
I have tested mostly the existing text-only chat and infill functionality.

ggerganov commented on tools/server/server.cpp:
```diff
@@ -3166,8 +3268,11 @@ struct server_context {
-                        for (int i = 0; i < slot.n_prompt_tokens; ++i) {
-                            common_sampler_accept(slot.smpl, prompt_tokens[i], false);
+                        for (size_t i = 0; i < slot.cache_tokens.size(); ++i) {
+                            llama_token id = slot.prompt_tokens[i];
```
Should this be:

```suggestion
                            llama_token id = slot.cache_tokens[i];
```

I am not sure. Maybe put a `GGML_ASSERT(slot.cache_tokens.size() == slot.prompt_tokens.size());` and if it ever triggers we can resolve it.

ngxson committed:
Apply suggestions from code review

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

ngxson committed:
rm can_be_detokenized

ngxson committed:
on prmpt processing done, assert cache_tokens.size

ngxson committed:
handle_completions_impl returns void

ngxson committed:
Merge branch 'master' into xsn/server_mtmd

ngxson committed:
adapt the new web ui

ngxson commented on tools/server/server.cpp:
```diff
@@ -4059,19 +4225,23 @@ int main(int argc, char ** argv) {
     const auto handle_completions = [&handle_completions_impl](const httplib::Request & req, httplib::Response & res) {
         json data = json::parse(req.body);
+        std::vector<raw_buffer> files; // dummy
         return handle_completions_impl(
```
I change `handle_completion_impl` to explicitly return void in https://github.com/ggml-org/llama.cpp/pull/12898/commits/b2906a98278e7ca532718800b4ac64553a78b16b

ngxson commented:
One more thing before merging this: I need to update the docs

ngxson committed:
update docs and hot topics

ngxson commented:
I updated mtmd docs to state that llama-server is supported, while also mentioning in llama-server docs that `/chat/completions` now support this.

Also added to hot topic. Merging this once the CI passes 🔥 

---

Here is the list of follow-up PR that I'll work on:
- Allow interrupt mid-processing the image - useful in case user wants to cancel the request. Discussion here: https://github.com/ggml-org/llama.cpp/pull/13366?notification_referrer_id=NT_kwDOAHWGu7MxNjI4MjA1MjMxMzo3NzAyMjAz#issuecomment-2866283804
- Support raw `/completions` API
- Allow caching image encoded embeddings, so same image won't need to be process again if it appears in 2 different chat history (not same prefix)
- Maybe support embedding model, discussion here: https://github.com/ggml-org/llama.cpp/issues/13247

ngxson commented on tools/server/server.cpp:
```diff
@@ -3140,34 +3206,70 @@ struct server_context {
 
                     // entire prompt has been processed
                     if (slot.n_past == slot.n_prompt_tokens) {
+                        GGML_ASSERT(slot.cache_tokens.size() == slot.prompt_tokens.size());
```
Hmm annoyingly, I'm running the test locally and this fails in the `unit/test_completion.py::test_consistent_result_same_seed`

ngxson replied:
@ggerganov it also fails on `master` so I think there is something wrong but not related to this PR, therefore, I'll comment out this assertion for now and leave a comment.

In any cases, I think `cache_tokens` and `prompt_tokens` can't guaranteed to be the same length if prompt truncation or context shifting being used

ggerganov replied:
On `master` it is different logic - it loops to `n_prompt_tokens`.

ggerganov replied:
On `master`, this assert is correct:

```diff
diff --git a/tools/server/server.cpp b/tools/server/server.cpp
index 06788bbdc..e7c75cd63 100644
--- a/tools/server/server.cpp
+++ b/tools/server/server.cpp
@@ -3166,6 +3166,7 @@ struct server_context {
                         common_sampler_reset(slot.smpl);
 
                         // Process all prompt tokens through sampler system
+                        GGML_ASSERT(slot.n_prompt_tokens == prompt_tokens.size());
                         for (int i = 0; i < slot.n_prompt_tokens; ++i) {
                             common_sampler_accept(slot.smpl, prompt_tokens[i], false);
                         }
```

So I think this change here introduces a bug - it's better to fix it.

Maybe this patch:

```diff
diff --git a/tools/server/server.cpp b/tools/server/server.cpp
index df12b0dab..de8ded71f 100644
--- a/tools/server/server.cpp
+++ b/tools/server/server.cpp
@@ -3256,18 +3256,16 @@ struct server_context {
 
                     // entire prompt has been processed
                     if (slot.n_past == slot.n_prompt_tokens) {
-                        // TODO @ngxson : this assertion fails sometimes, why?
-                        // GGML_ASSERT(slot.cache_tokens.size() == slot.prompt_tokens.size());
-
                         slot.state = SLOT_STATE_DONE_PROMPT;
 
                         GGML_ASSERT(batch.n_tokens > 0);
+                        GGML_ASSERT((size_t) slot.n_prompt_tokens == slot.prompt_tokens.size());
 
                         common_sampler_reset(slot.smpl);
 
                         // Process all prompt tokens through sampler system
-                        for (size_t i = 0; i < slot.cache_tokens.size(); ++i) {
-                            llama_token id = slot.cache_tokens[i];
+                        for (int i = 0; i < slot.n_prompt_tokens; ++i) {
+                            llama_token id = slot.prompt_tokens[i];
                             if (id != LLAMA_TOKEN_NULL) {
                                 common_sampler_accept(slot.smpl, id, false);
                             }

```

ngxson replied:
Thanks, that fixes the issue, done in https://github.com/ggml-org/llama.cpp/pull/12898/commits/9ed430ca267176f73c407302be36aa431a82e352

ngxson committed:
rm assert

ngxson committed:
small fix (2)

broadbit-hu commented:
Something went wrong since this commit with long input token counts (possibly greater than 2048):
```
main: server is listening on http://0.0.0.0:18080 - starting the main loop
srv  update_slots: all slots are idle
srv  log_server_r: request: GET /props 192.168.253.130 200
slot launch_slot_: id  0 | task 0 | processing task
slot update_slots: id  0 | task 0 | new prompt, n_ctx_slot = 22016, n_keep = 0, n_prompt_tokens = 8241
slot update_slots: id  0 | task 0 | kv cache rm [0, end)
slot update_slots: id  0 | task 0 | prompt processing progress, n_past = 2048, n_tokens = 2048, progress = 0.248514
/opt/text/llama.cpp/tools/server/utils.hpp:1157: GGML_ASSERT(n <= tokens.size()) failedslot update_slots: id  0 | task 0 | kv cache rm [2048, end)

Memory critical error by agent node-0 (Agent handle: 0x59fc5fabc930) on address 0x7cbd6cc00000. Reason: Memory in use. 
Aborted (core dumped)
```

Same results on ROCm and NVidia CUDA too...

Bug ticket: [#13484](https://github.com/ggml-org/llama.cpp/issues/13484)

LordLokator commented:
@ngxson  Very well done, thank you for this!

aviallon commented:
> @ngxson  Very well done, thank you for this! 

@ngxson tested in production. Speed, quality and ease-of-use is impressive. I simply had to update my llama.cpp build, and it _just works_.
Major accomplishment. Well done.
